[3D] Clipping planes tool

Withalion opened the pull request:
## Description

This PR adds a new tool to create cross sections in 3D scene. User enables this tool in 3D window, selects the area on 2D canvas and after confirmation the 3D scene will be edited + the camera will change to profile view.
`Left click` -> adds point
~~`BACKSPACE` or `DEL` -> remove last step in selection~~
`Right click` or `ESCAPE`  -> resets the selection


https://github.com/user-attachments/assets/3e0e7a67-d28d-48d4-a342-ed267f2d603a



Funded by: Septima, Klimadatastyrelsen


<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

Withalion committed:
Add tool to set clipping planes in 3D scene

Withalion committed:
Add perpendicular & parallel clipping planes

Withalion committed:
Implement support for 4 clipping planes

Withalion committed:
Refactor & fix some stuff

Withalion committed:
Refactor clipping tool to set view from profile after clip

Withalion committed:
Refactor clipping tool again & improve tests

nyalldawson commented:
Why does this require multiple attempts in the screencast?

Withalion commented:
I see that I won't be a youtuber... I was trying to show that users can undo/remove the selection

Withalion committed:
Fix test to compile for qt 6

wonder-sk commented on tests/src/3d/testqgs3dutils.cpp:
```diff
@@ -516,6 +519,88 @@ void TestQgs3DUtils::testScreenPointToMapCoordinates()
+  const QgsPointXY point2( 50, 50 );
+  const QgsVector3D midPoint( point2.x() - point1.x(), point2.y() - point1.y(), 0 );
+  QList<QVector4D> clippingPlanes = Qgs3DUtils::lineSegmentToClippingPlanes( point1, point2, 10, QgsVector3D( 0, 0, 0 ) );
+  for ( int i = 0; i < clippingPlanes.size(); i++ )
+  {
+    QgsVector3D planePoint(
+      i % 2 ? midPoint.x() : ( -clippingPlanes.at( i ).y() * midPoint.y() - clippingPlanes.at( i ).w() ) / clippingPlanes.at( i ).x(),
+      i % 2 ? ( -clippingPlanes.at( i ).x() * midPoint.x() - clippingPlanes.at( i ).w() ) / clippingPlanes.at( i ).y() : midPoint.y(),
+      0
+    );
+    const double distance = QgsVector3D::dotProduct( midPoint - planePoint, clippingPlanes.at( i ).toVector3D() );
+    // verify all normals are pointing inside the rectangle
+    QVERIFY( distance > 0 );
+  }
```
let's check we get exactly 4 clipping planes as a result...

and let's simplify this - pick just one point that's clearly inside (e.g. 30, 30) and check that it is inside (i.e. all dot products are positive).

and let's also take a point that's outside and check that at least one dot product is negative...

wonder-sk commented on tests/src/3d/testqgs3dutils.cpp:
```diff
@@ -516,6 +519,88 @@ void TestQgs3DUtils::testScreenPointToMapCoordinates()
+    QVERIFY( distance > 0 );
+  }
+
+  QVERIFY( QVector4D::dotProduct( clippingPlanes.at( 0 ).toVector3D(), clippingPlanes.at( 1 ).toVector3D() ) == 0 );
+  QVERIFY( QVector4D::dotProduct( clippingPlanes.at( 1 ).toVector3D(), clippingPlanes.at( 2 ).toVector3D() ) == 0 );
+  QVERIFY( QVector4D::dotProduct( clippingPlanes.at( 2 ).toVector3D(), clippingPlanes.at( 3 ).toVector3D() ) == 0 );
+  QVERIFY( QVector4D::dotProduct( clippingPlanes.at( 3 ).toVector3D(), clippingPlanes.at( 0 ).toVector3D() ) == 0 );
```
I would probably remove these checks - there's a good chance they will break due to some numerical issues and return some small number instead

wonder-sk commented on tests/src/3d/testqgs3dutils.cpp:
```diff
@@ -516,6 +519,88 @@ void TestQgs3DUtils::testScreenPointToMapCoordinates()
+    const double distance = QgsVector3D::dotProduct( midPoint - planePoint, clippingPlanes.at( i ).toVector3D() );
+    // verify all normals are pointing inside the rectangle
+    QVERIFY( distance > 0 );
+  }
```
to limit the amount of duplicate code, it may be useful to build a utility function (e.g. isInsideClippingPlanes()) for testing

wonder-sk reviewed (commented):
Clipping in 3D - yay! :tada:

nyalldawson commented:
@Withalion 
>  I see that I won't be a youtuber... I was trying to show that users can undo/remove the selection

Ah, makes sense! Can you remove the incorrect warning about "no vector layer selected" which is shown when undoing?

Withalion commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,158 @@
+      mRubberBandPolygon->reset( Qgis::GeometryType::Polygon );
+    }
+  }
+  //TODO: somehow consume the event so we don't get the annoying info message about not having vector layer
```
So `QgsMapCanvas` consume the event sooner than the map tool, for now I'll remove the the possibility for undo during selection and just leave the cancellation with `ESC`

Withalion commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,158 @@
+        mRectangleWidth,
+        mSceneOrigin
+      );
+      emit clippingPlanesChanged( clippingPlanes );
```
Now because of the camera reposition we still need to pass the clipping planes anyway...

ptitjano replied:
I agree. You can add the appropriate getter, if necessary.

Withalion committed:
Fix review issues + add some goodies

After cross section is selected camera looks eastward at the selected
cross section from distance to see it whole in profile view. User also
sees highlighted area of cross section in 2D canvas.

Withalion committed:
Fix camera issues in 3D after clipping scene

ptitjano commented on src/3d/materials/qgsmaterial.cpp:
```diff
@@ -120,7 +120,6 @@ void QgsMaterial::disableClipping()
     if ( parameterName == QgsMaterial::CLIP_PLANE_ARRAY_PARAMETER_NAME || parameterName == QgsMaterial::CLIP_PLANE_MAX_PLANE_PARAMETER_NAME )
     {
       materialEffect->removeParameter( parameter );
-      break;
```
Good catch. I have created https://github.com/qgis/QGIS/pull/61135 for backport

Withalion replied:
Thanks for that!

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+/***************************************************************************
+    qgsmaptoolclippingplanes.h  -
```
qgsmaptoolclippingplanes.cpp

ptitjano commented on tests/src/3d/testqgs3dutils.cpp:
```diff
@@ -39,6 +39,15 @@ static bool qgsQuaternionNear( const QQuaternion &q1, const QQuaternion &q2, dou
 }
 
+static bool isPointInFrontOfPlane( const QgsVector3D &point, const QgsVector3D &planePoint, const QgsVector3D &planeNormal )
+{
+  if ( QgsVector3D::dotProduct( point - planePoint, planeNormal ) > 0 )
+  {
+    return true;
+  }
+  return false;
+}
+
```
return ( QgsVector3D::dotProduct( point - planePoint, planeNormal ) > 0 );

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+    const QVector<QgsPointXY> points( {
+      *mRubberBandPoints->getPoint( 0, 0 ) + vec * mRectangleWidth,  //! top left corner
+      *mRubberBandPoints->getPoint( 0, 1 ) + vec * mRectangleWidth,  //! top right corner
+      *mRubberBandPoints->getPoint( 0, 1 ) + -vec * mRectangleWidth, //! bottom right corner
+      *mRubberBandPoints->getPoint( 0, 0 ) + -vec * mRectangleWidth  //! bottom left corner
```
`+ -` => `-`

ptitjano commented on src/3d/qgs3dutils.cpp:
```diff
@@ -1026,3 +1026,38 @@ QgsPoint Qgs3DUtils::screenPointToMapCoordinates( const QPoint &screenPoint, con
   return mapPoint;
 }
+
+QList<QVector4D> Qgs3DUtils::lineSegmentToClippingPlanes( const QgsVector3D &point1, const QgsVector3D &point2, const double distance, const QgsVector3D &origin )
```
According to the docstring, the `distance` is supposed to be positive. This should be checked at the beginning of the function.

ptitjano commented on src/3d/qgs3dutils.cpp:
```diff
@@ -1026,3 +1026,38 @@ QgsPoint Qgs3DUtils::screenPointToMapCoordinates( const QPoint &screenPoint, con
+{
+  QgsVector3D lineDirection( point2 - point1 );
+  lineDirection.normalize();
+  const QgsVector tempVec( lineDirection.x(), lineDirection.y() );
```
`tempVec` => `lineDirection2D`

ptitjano replied:
even better: `const QgsVector lineDirection2DPerp = QgsVector( lineDirection.x(), lineDirection.y() ).perpVector();

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1178,67 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+  // check if the angle between the view point is to the left or right of the scene north, apply angle offset if necessary for camera
+  if ( QgsVector3D::crossProduct( planes.at( 1 ).toVector3D(), QgsVector3D( 0, -1, 0 ) ).z() > 0 )
+  {
+    yawAngle = 180 + ( 180 - yawAngle );
```
yawAngle = 360 - yawAngle;

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+  else if ( e->button() == Qt::RightButton )
+  {
+    clearRubberBand();
+    mRubberBandPolygon->reset( Qgis::GeometryType::Polygon );
```
clearHighLightedArea();

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+{
+  if ( mRubberBandPoints->numberOfVertices() == 0 )
+  {
+    mRubberBandPolygon->reset( Qgis::GeometryType::Polygon );
```
clearHighLightedArea();

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+  if ( e->key() == Qt::Key_Escape )
+  {
+    clearRubberBand();
+    mRubberBandPolygon->reset( Qgis::GeometryType::Polygon );
```
clearHighLightedArea();

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+void QgsMapToolClippingPlanes::clearRubberBand() const
+{
+  mRubberBandLines->reset( Qgis::GeometryType::Line );
+  mRubberBandPoints->reset( Qgis::GeometryType::Point );
+}
+
+void QgsMapToolClippingPlanes::clearHighLightedArea() const
+{
+  mRubberBandPolygon->reset( Qgis::GeometryType::Polygon );
+}
```
Add a `clear` which calls `clearRubberband` and `clearHighLightedArea`

ptitjano commented on src/3d/qgs3dutils.h:
```diff
@@ -363,6 +363,16 @@ class _3D_EXPORT Qgs3DUtils
+     * \note \a distance is expected to be positive
+     * \since QGIS 3.44
+     */
+    static QList<QVector4D> lineSegmentToClippingPlanes( const QgsVector3D &point1, const QgsVector3D &point2, double distance, const QgsVector3D &origin );
```
Please rename `point1` to `startPoint` and `point2` to `endPoint`

ptitjano commented on src/3d/qgs3dutils.h:
```diff
@@ -363,6 +363,16 @@ class _3D_EXPORT Qgs3DUtils
     static QgsPoint screenPointToMapCoordinates( const QPoint &screenPoint, QSize size, const QgsCameraController *cameraController, const Qgs3DMapSettings *mapSettings );
+
+    /**
+     * Returns list of 4 planes, which are calculated from line defined by \a point1 and \a point2. \a distance defines
+     * the distance between parallel clipping planes and the line.
+     * Clipping planes are returned back in the format of 4D Vector, where first 3 coordinates are the coordinates of
+     * normalized normal of given plane and the last is the distance from origin of the scene.
```
I find the wording hard to understand. Could you please rephrase it. Maybe something like: 

```
Returns a list of 4 planes derived from a line extending from \a startPoint to \a endPoint. The parameter \a distance defines the distance between the parallel clipping planes and the line.

Each clipping plane is represented as a 4D vector, where the first three components correspond to the normalized normal of the plane, and the fourth component represents its distance from the origin of the scene.
```

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1178,67 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+  mMainCanvas->setMapTool( mMapToolClippingPlanes.get() );
+  QgisApp::instance()->activateWindow();
+  QgisApp::instance()->raise();
+  mMessageBar->pushInfo( QString(), tr( "Select a rectangle by 3 points on the main 2D map view to define this 3D scene's cross section" ) );
```
`Select a rectangle using 3 points on the main 2D map view to define the cross-section of this 3D scene`

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1178,67 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+
+  // calculate the middle of the front side defined by clipping planes with Cramer's rule
+  double det = planes.at( 0 ).x() * planes.at( 3 ).y() - planes.at( 3 ).x() * planes.at( 0 ).y();
+  const QgsVector3D point1(
+    ( planes.at( 0 ).w() * planes.at( 3 ).y() - planes.at( 3 ).w() * planes.at( 0 ).y() ) / -det,
+    ( planes.at( 3 ).w() * planes.at( 0 ).x() - planes.at( 0 ).w() * planes.at( 3 ).x() ) / -det,
+    0
+  );
+  det = planes.at( 2 ).x() * planes.at( 3 ).y() - planes.at( 3 ).x() * planes.at( 2 ).y();
+  const QgsVector3D point2(
+    ( planes.at( 2 ).w() * planes.at( 3 ).y() - planes.at( 3 ).w() * planes.at( 2 ).y() ) / -det,
+    ( planes.at( 3 ).w() * planes.at( 2 ).x() - planes.at( 2 ).w() * planes.at( 3 ).x() ) / -det,
+    0
+  );
```
could you please rename `point1` to `startPoint`, `point2` to `endPoint` and create variables for `planes.at( 0 )`,  `planes.at( 2 )`,  `planes.at( 3 )`. 

This will improve readability.

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1178,67 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+  );
+  QgsVector3D middle( point1 + ( point2 - point1 ) / 2 );
+  // we temporarily disable terrain so the elevation calculation won't take it in account if the terrain is flat
+  const bool lastTerrainState = mCanvas->scene()->mapSettings()->terrainRenderingEnabled();
+  if ( dynamic_cast<const QgsFlatTerrainSettings *>( mCanvas->scene()->mapSettings()->terrainSettings() ) )
+  {
+    mCanvas->scene()->mapSettings()->setTerrainRenderingEnabled( false );
+  }
+  const double side = std::max( middle.distance( point1 ), ( mCanvas->scene()->elevationRange().upper() - mCanvas->scene()->elevationRange().lower() ) / 2 );
```
I think that It should be possible to add a `ignoreTerrain` parameter to `mCanvas->scene()->elevationRange()` to ignore the terrain and avoid disabling the terrain here.

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1178,67 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+  {
+    yawAngle = 180 + ( 180 - yawAngle );
+  }
+  mCanvas->scene()->elevationRange().isInfinite() ? middle.setZ( 0 ) : middle.setZ( mCanvas->scene()->elevationRange().lower() + ( mCanvas->scene()->elevationRange().upper() - mCanvas->scene()->elevationRange().lower() ) / 2 );
```
`mCanvas->scene()->elevationRange()` is used multiple times, please store it in a variable.

Withalion committed:
Fix review issues

ptitjano commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -1103,11 +1103,11 @@ QgsRectangle Qgs3DMapScene::sceneExtent() const
 }
 
-QgsDoubleRange Qgs3DMapScene::elevationRange() const
+QgsDoubleRange Qgs3DMapScene::elevationRange( const bool ignoreTerrain ) const
```
Could you please add a unit test which tests both cases?

Withalion committed:
Refactor code & add unit tests

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,154 @@
+
+void QgsMapToolClippingPlanes::deactivate()
+{
+  clearRubberBand();
```
why not calling `clear()` here?

Withalion replied:
Because we want to highlight the cross-section on main canvas after the user is done with the selection.

ptitjano commented on src/3d/qgs3dutils.cpp:
```diff
@@ -1026,3 +1026,73 @@ QgsPoint Qgs3DUtils::screenPointToMapCoordinates( const QPoint &screenPoint, con
+  QgsCameraPose cameraPose;
+  cameraPose.setPitchAngle( 90 );
+
+  // // calculate the middle of the front side defined by clipping planes
```
No need for double comments here

ptitjano commented on src/3d/qgs3dutils.h:
```diff
@@ -363,6 +365,24 @@ class _3D_EXPORT Qgs3DUtils
+    static QList<QVector4D> lineSegmentToClippingPlanes( const QgsVector3D &startPoint, const QgsVector3D &endPoint, double distance, const QgsVector3D &origin );
+
+    /**
+     * Returns camera pose of camera, which is looking at mid-point between \a startPoint and \a endPoint. The camera
+     * is tilted to look from profile and the heading angle is calculated to look from the right side of the line.
+     * The middle of \a elevationRange sets the z attribute of mid-point.
```
Could you please rephrase it a little bit?
Maybe, someting like: 

```
Returns the camera pose for a camera looking at mid-point between \a startPoint and \a endPoint. 
The camera is angled to provide a profile view and the heading angle is calculated to look from the right side of the line.
```

ptitjano commented on tests/src/3d/testqgs3dutils.cpp:
```diff
@@ -516,6 +535,67 @@ void TestQgs3DUtils::testScreenPointToMapCoordinates()
+  QCOMPARE( camPose.headingAngle(), 45 );
+  QGSCOMPARENEAR( camPose.distanceFromCenterPoint(), 46, 0.2 );
+
+  // test 1: the distance between start point and end point is smaller than elevation range
```
typo: "test 2"

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.h:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * \ingroup app
+ * \brief A map tool that stores clipping planes from lines drawn onto map canvas.
```
nitpick: this tool does not "store" the clipping planes, it allows to "define" them.

Withalion committed:
Fix final review issues

ptitjano commented:
I just did some testing on a 3D view which only contains Vector layers. The tools works. However, the zoom is not correctly updated. I end up with a 3d scene which looks empty. But it's not. If I click on the "reset view" button, the scene is correctly computed. See the video below.

I think this is because `Qgs3DMapScene::elevationRange` is not able to able vector layers. In that case (if the elevation range cannot be computed), we should fallback to the defaut view by calling  `Qgs3DMapScene::viewZoomFull` to avoid a 3d scene which seems "broken" or empty.


https://github.com/user-attachments/assets/cd837ec8-6615-4602-bee8-ca08aabe39ed

ptitjano commented:
I also tried to define clipping planes outside of the scene. In that case, we end up with an empty 3d scene. I'm not sure how to handle this situation. In that case, I would propose : 
- do not set the clipping planes
- display an error message to explain the situation

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -1168,6 +1177,58 @@ void Qgs3DMapCanvasWidget::setSceneExtent( const QgsRectangle &extent )
+  mMessageBar->pushInfo( QString(), tr( "Select a rectangle using 3 points on the main 2D map view to define the cross-section of this 3D scene" ) );
+}
+
+void Qgs3DMapCanvasWidget::setViewOnClippingPlanesChanged( const QgsVector3D &startPoint, const QgsVector3D &endPoint, const double width )
```
It looks like you changed from a `slot` to a function but you haven't changed its name. The current name does not make a lot of sense to me. 

I would prefer something like `enableClippingPlanes`. In that case, it would take 2 arguments: the list of clipping planes and the new camera. Which means that the clipping planes and the new QgsCameraPose should be computed in `QgsMapToolClippingPlanes`.

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -258,6 +260,10 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
 
+  mActionSetClippingPlanes = mCameraMenu->addAction( QgsApplication::getThemeIcon( QStringLiteral( "mActionEditCut.svg" ) ), tr( "Cross Section Tool" ), this, &Qgs3DMapCanvasWidget::setClippingPlanesOn2DCanvas );
+  mActionSetClippingPlanes->setCheckable( true );
+  mCameraMenu->addAction( QgsApplication::getThemeIcon( QStringLiteral( "mActionEditCutDisabled.svg" ) ), tr( "Disable Cross Section" ), this, &Qgs3DMapCanvasWidget::disableClippingPlanes );
```
This action should be disabled if no clipping plane is defined.

ptitjano commented:
> @ptitjano please let me know when you are done with the review so I can push the changes

Thank you for taking into account my remarks. I did some testing and it seems to work reliably. Great! However, I've noticed some corner cases (see my previous comments and my video) and I'm not sure how this tool should work with the existing extent definition from the configuration dialog.

wonder-sk commented:
> I don't understand how this new tool is supposed to work with the extent defined in the 3D configuration widget. I'm not sure what the expected behavior should be.

@ptitjano the extent defined in the 3D configuration widget should be treated as a kind of "project extent" - something that users would normally set once or not set it at all, and just keep the default extent. That's the area of interest for the whole 3D scene. The clipping planes are meant for temporary clipping of the 3D scene, to inspect some details, to get a better idea of a cross-section in a particular location etc.

The big difference is that when 3D scene extent is changed, this forces a complete reload of the 3D scene - something that should be minimized as much as possible. The clipping planes approach is very light, as it does not require full scene reloads, at the same time it should not replace the extent setup.

> -    when a cross-section is defined, should the extent in the configuration be updated?

No, the extent should stay the same.

> -    when the clipping planes are defined, if I update the scene extent from the configuration widget, should the clipping planes be updated ?

Not really. But I can imagine a behavior where clipping would get disabled, if the new scene extent has no intersection with the existing clip planes (and thus the scene would end up empty).

Withalion commented:
Hello @ptitjano, I addressed the issues you pointed out.
> I just did some testing on a 3D view which only contains Vector layers. The tools works. However, the zoom is not correctly updated. I end up with a 3d scene which looks empty.

There was actually a bug regarding the layers for which we don't compute elevation range. That should be fixed now and you should get correct profile view. I decided to not revert back to `Qgs3DMapScene::viewZoomFull`, users still get the profile view, but slightly tilted so even if you have just flat layers you'll see them in the scene.

> I also tried to define clipping planes outside of the scene.

As you proposed users will get notification in the message bar that no cross-section has been selected, why not and that they should try again. Also the tool will react accordingly...

And I see that your concerns about the extent tool and cross-section tool were already answered by @wonder-sk. Also regarding 
> But I can imagine a behavior where clipping would get disabled, if the new scene extent has no intersection with the existing clip planes (and thus the scene would end up empty).

First we will have to fix the issue with origin shift. Because setting new extent also moves the origin, so currently there will be always intersection.

ptitjano commented:
@Withalion @wonder-sk Thanks for the explanation. However, I have just tried the new version and I still have a few usability concerns

1. The colors between the "scene extent" and the "cross section tool" are too similar in the 2D canvas

See my screenshot. I've activated the whole scene extent from the configuration and enabled a cross section. Both rectangles are red. It is hard to distinguish them and to define the cross-section with the existing red rectangle from the 3D scene extent

![Capture d’écran du 2025-03-26 12-07-01](https://github.com/user-attachments/assets/28de1ae1-1cbc-4a7b-baa0-b86a9bc6af3e)

2. In the 3D view, It is not obvious to understand if the cross section tool is enabled

Yes, The cross-section rectangle is visible in the 2D canvas but if only look at the 3d view, I don't know easily if the cross sections are enabled.
For example, in the effects menu, if the shadows are enabled, a checkbox is clicked. I can easily check if the effect is activated, activate or deactivate it.
In the cross section case, I find it less obvious. I can understand why there are 2 different entries in the menu. You may want to change an existing cross-section. But it makes it harder to understand if the cross-sections are enabled.
To sum it up, I think the 3D view is lacking some kind of visual indication that the cross-section are active

3. I create a new 3D scene, I define clipping planes outside of the 3D scene extent. The clipping planes are enabled and I end up with an empty view

Based on your recent comments, I've understood that now the clipping planes should not be enabled and a message should be displayed to warn the user. I have not been able to reproduce this behavior. I always end up with an empty view.

4. I create a new 3D scene. I define clipping planes in the scene. It works. Then I change the 3D scene extent with "Set 3D extent on 2D map view" and select a new extent outside of the clipping planes. I end up with an empty view.

Based on your recent comments, I understand that this case won't be handled by this PR. Is that correct?

Withalion commented:
> 1. The colors between the "scene extent" and the "cross section tool" are too similar in the 2D canvas

I don't from where is picking color the extent rectangle, however the cross-section area highlight rectangle has the same color as the highlight color. Users can change that in options.

> 2. In the 3D view, It is not obvious to understand if the cross section tool is enabled

Unfortunately the `QAction` in the camera menu that user clicks can only show icon or check mark so there is no easy way to show it there. I think we can leave it to the users intuition for now, so when I can disable the clip planes that means they are active. Plus there is always the indication on the 2D canvas and I expect most of the users have both open. Maybe we can leave this for a future PR to improve the UX of this tool...

> 3. I create a new 3D scene, I define clipping planes outside of the 3D scene extent. The clipping planes are enabled and I end up with an empty view

I was not able to reproduce this issue. It takes all visible layers into account. Maybe it looks like it's outside of the extents, but it's not? Also the clipping planes have to be completely outside to get this notification. If it keeps happening please post a screencast of the bug.

> 4. I create a new 3D scene. I define clipping planes in the scene. It works. Then I change the 3D scene extent with "Set 3D extent on 2D map view" and select a new extent outside of the clipping planes. I end up with an empty view.

Yes, you are correct

Withalion commented:
It's not necessary as the PR with fix will come soon. (today/tomorrow)

Also I'll try to fix up this PR today so we can move forward.

ptitjano commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -0,0 +1,188 @@
+      {
+        clear();
+        QgsMessageBar *msgBar = QgisApp::instance()->messageBar();
+        msgBar->pushInfo( QString(), tr( "The cross section is outside of the scene extent, please select a new one!" ) );
```
nitpick. No need for a `of` here: `The cross section is outside the scene extent, please select a new one!`

ptitjano commented:
Thanks @Withalion I did more testing and now I cannot define a cross-sections outside the scene.

Regarding my other issue, I'm afraid I haven't understood your answer. If the cross-sections are already active and I change the 3D scene extent outside the cross-sections extent, I end up with an empty view. Is it supposed to work and it won't be covered by this PR?

Withalion commented:
Anything related to changing extents, which also means moving the origin of the scene, is not covered by this PR. It will be covered by another one. Take this as a base PR for the clipping tool and we will hammer out any bugs with next PRs.

ptitjano commented:
> Anything related to changing extents, which also means moving the origin of the scene, is not covered by this PR. It will be covered by another one. Take this as a base PR for the clipping tool and we will hammer out any bugs with next PRs.

In that case, update this PR once the handling of the origin of the scene is ready. This will ensure to merge a version without known issues or limitation.

wonder-sk committed:
Support for rebasing of scenes

wonder-sk commented:
Handling of origin rebasing added in https://github.com/qgis/QGIS/pull/61057/commits/bb14eabc85746735c1c1b50ef941acb869646342

ptitjano commented:
@Withalion @wonder-sk I have just tried an up to date QGIS master and I can still reproduce the issue previously mentioned. 


https://github.com/user-attachments/assets/8a10bfbf-6cef-40c2-acba-a65d152de1da


Whether or not this issue would have been fixed, I also want to point out that I find it somewhat disrespectful to merge a PR without asking for my opinion, especially since I spent a lot of time reviewing and testing it.

wonder-sk commented:
> Whether or not this issue would have been fixed, I also want to point out that I find it somewhat disrespectful to merge a PR without asking for my opinion, especially since I spent a lot of time reviewing and testing it.

@ptitjano Understood. We didn't intend any disrespect, and apologize if it came across that way. We felt the PR was ready and didn't think you had further feedback, but we should have checked in. We'll be more mindful in the future.

I am sure @Withalion will check your bug and fix if it can be replicated.
